http/src/main/java/io/micronaut/http/ssl/SslBuilder.java

```diff
@@ -137,33 +152,269 @@ protected Optional<KeyStore> getKeyStore(SslConfiguration ssl) throws Exception
+                            store = createEmptyKeyStore(spec.provider, "PKCS12");
+                        }
+
+                        loadPem(ks.path, spec.password, spec.provider, store);
+                    } catch (PemParser.NotPemException f) {
+                        // probably should have been loaded as KS
+                        e.addSuppressed(new Exception("Also tried and failed to load the input as PEM:", f));
+                        throw e;
+                    } catch (Exception f) {
+                        // probably should have been loaded as PEM
+                        f.addSuppressed(new Exception("Also tried and failed to load the input as a key store:", e));
+                        throw f;
+                    }
+                } else {
+                    throw e;
+                }
+            }
             return store;
+        } else if (spec instanceof PemBasedCertificateSpec pem) {
+            List<Object> keyItems;
+            try (InputStream s = resourceResolver.getResourceAsStream(pem.keyPath).orElseThrow()) {
+                keyItems = new PemParser(pem.provider, pem.password)
+                    .loadPem(new String(s.readAllBytes(), StandardCharsets.UTF_8));
+            }
+            List<Object> certItems;
+            try (InputStream s = resourceResolver.getResourceAsStream(pem.certificatePath).orElseThrow()) {
+                certItems = new PemParser(pem.provider, pem.password)
+                    .loadPem(new String(s.readAllBytes(), StandardCharsets.UTF_8));
+            }
+
+            if (keyItems.size() != 1) {
+                throw new IllegalArgumentException("key-path contained more than one PEM object. It should only contain the private key.");
+            }
+            if (!(keyItems.get(0) instanceof PrivateKey pk)) {
+                throw new IllegalArgumentException("key-path contained a certificate instead of a private key.");
+            }
+            KeyStore store = createEmptyKeyStore(spec.provider, spec.type == null ? "PKCS12" : spec.type);
+            store.load(null, null);
+            store.setKeyEntry(
+                "key",
+                pk,
+                null,
+                certificates(certItems).toArray(new X509Certificate[0])
+            );
+            return store;
+        } else {
+            throw new AssertionError("Weird CertificateSpec");
+        }
+    }
+
+    private static @NonNull KeyStore createEmptyKeyStore(@Nullable String provider, String type) throws KeyStoreException, NoSuchProviderException {
+        return provider == null ? KeyStore.getInstance(type) : KeyStore.getInstance(type, provider);
+    }
+
+    private void loadPem(@NonNull String resource, @Nullable String password, @Nullable String provider, KeyStore store) throws IOException, GeneralSecurityException, PemParser.NotPemException {
+        List<Object> items;
+        try (InputStream s = resourceResolver.getResourceAsStream(resource).orElseThrow()) {
```

Comment:
throw a better exception here that includes a path to the resource that failed to load